src/Symfony/Component/TypeInfo/Type/CollectionType.php

```diff
@@ -92,6 +92,39 @@ public function wrappedTypeIsSatisfiedBy(callable $specification): bool
         return $this->getWrappedType()->isSatisfiedBy($specification);
     }
 
+    public function accepts(mixed $value): bool
+    {
+        if (!parent::accepts($value)) {
+            return false;
+        }
+
+        // array is not a list
+        if ($this->isList() && \is_array($value) && !array_is_list($value)) {
+            return false;
+        }
+
+        $keyType = $this->getCollectionKeyType();
+        $valueType = $this->getCollectionValueType();
+
+        $i = 0;
+
+        foreach ($value as $k => $v) {
+            // key or value do not match
+            if (!$keyType->accepts($k) || !$valueType->accepts($v)) {
+                return false;
+            }
+
+            // iterable is not a list
+            if ($this->isList() && !\is_array($value) && $i !== $k) {
+                return false;
+            }
```

Comment:
Noted this condition is for when `$value` is an Iterator and the type is a list. For `array`, this is already checked quickly with `array_is_list`.

Reply:
a `list<T>` does not accept iterables in phpstan or Psalm. It is a subtype of `array<T>`. I think we should keep it consistent. The case of `$this->isList() && !\is_array($value)` should return early, before the iteration.

Reply:
Updated.

Maybe `CollectionType::__construct` must be updated as well to deprecate (then prevent) the creation of an "iterable list", WDYT?

Reply:
Here you go: https://github.com/symfony/symfony/pull/59302